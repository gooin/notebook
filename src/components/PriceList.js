import React from 'react';
import {Row, Col, Tooltip, Button, Divider} from 'antd';
import {EditOutlined, DeleteOutlined} from '@ant-design/icons';
import Icons from './Icons';
import PropTypes from 'prop-types';

const PriceList = ({items, onEdit, onDelete}) => {
    console.log('items :>> ', items);
    return (
        <>
            {
                items.map((item) => (
                    <div key={item.id}>
                        <Divider/>
                        <Row>
                            <Col span={4}>
                                <Icons iconName={item.category.name}/>
                            </Col>
                            <Col span={8}>{item.title}</Col>
                            <Col span={4}>
                                {item.category.type === "outcome" ? "-" : "+"}
                                {item.price}</Col>
                            <Col span={6}>{item.date}</Col>
                            <Col span={1}>
                                <Tooltip title="编辑">
                                    <Button type="primary" size="small" shape="circle" icon={<EditOutlined/>}
                                            onClick={() => {
                                                onEdit(item)
                                            }}
                                    />
                                </Tooltip>
                            </Col>
                            <Col span={1}>
                                <Tooltip title="删除">
                                    <Button type="primary" size="small" danger shape="circle" icon={<DeleteOutlined/>}
                                            onClick={() => {
                                                onDelete(item)
                                            }}
                                    />
                                </Tooltip>
                            </Col>
                        </Row>

                    </div>
                ))
            }

        </>
    )
}

PriceList.propTypes = {
    items: PropTypes.array.isRequired,
    onEdit: PropTypes.func.isRequired,
    onDelete: PropTypes.func.isRequired
}
PriceList.defaultProps = {
    onEdit: () => {
        console.log("xxxxx")
    }
}

export default PriceList;