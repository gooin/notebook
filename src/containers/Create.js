import React, {Component} from 'react';

class Create extends Component {
    constructor(props) {
        super(props);
        this.state = {
            xx: 1
        }
    }

    render() {
        const { match} = this.props
        return (
            <div>
                <h1>createÔºÅ{match.params.id}</h1>
            </div>
        );
    }
}

export default Create;