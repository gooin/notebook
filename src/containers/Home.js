import React, {Component} from 'react';
import PriceList from '../components/PriceList'
import ViewTab from '../components/ViewTab'
import MonthPicker from '../components/MonthPicker'
import {Row, Col, Divider, Button} from 'antd';

import {LIST_VIEW, TYPE_INCOME, TYPE_OUTCOME, parseToYearAndMonth, CHART_VIEW} from '../utility'
import logo from "../logo.svg";
import TotalPrice from "../components/TotalPrice";
import CreateBtn from "../components/CreateBtn";


const items = [
    {
        "id": 1,
        "title": "吃饭",
        "price": 200,
        "date": "2020-11-12",
        "cid": 2
    },
    {
        "id": 2,
        "title": "假如发财了",
        "price": 200,
        "date": "2020-11-12",
        "cid": 3
    },
    {
        "id": 2,
        "title": "去去医院体检",
        "price": 200,
        "date": "2020-10-12",
        "cid": 1
    },{
        "id": 3,
        "title": "去去医院体检",
        "price": 200,
        "date": "2020-9-12",
        "cid": 1
    }
]

const categries = {
    "1": {
        "id": 1,
        "name": "travel",
        "type": "outcome"
    },
    "2": {
        "id": 2,
        "name": "food",
        "type": "outcome"
    },
    "3": {
        "id": 3,
        "name": "salary",
        "type": "income"
    }
}


const newItem = {
    "id": 5,
    "title": "工资",
    "price": 2000,
    "date": "2020-11-12",
    "cid": 3
}

class Home extends Component {
    constructor(props) {
        super(props);
        this.state = {
            items,
            currentDate: parseToYearAndMonth(),
            tabView: LIST_VIEW
        }
    }

    changeView = (type) => {
        console.log("changeView ", type);

        this.setState({
            tabView: type
        })
    }
    changeDate = (year, month) => {
        this.setState({
            currentDate: {
                year, month
            }
        })
    }
    editItem = (editedItem) => {
        // editedItem.title += "_updated! "
        const filtered = this.state.items.map(item => {
            if (editedItem.id === item.id) {
                return {...item, title: item.title + "_updated "}
            } else {
                return item
            }
        })
        this.setState({
            items: filtered
        })
    }
    deleteItem = (deletedItem) => {
        const filtered = this.state.items.filter(item => item.id !== deletedItem.id)
        this.setState({
            items: filtered
        })
    }
    createItem = () => {
        this.setState({
            items: [newItem, ...this.state.items]
        })
    }


    render() {
        const {items, currentDate, tabView} = this.state
        const itemsWithCategory = items.map(item => {
            item.category = categries[item.cid]
            return item
        }).filter(item => {
            // 过滤，仅显示当前年月的记录
            return item.date.includes(`${currentDate.year}-${currentDate.month}`)
        })

        let currIncome = 0;
        let currOutcome = 0;
        itemsWithCategory.forEach(item => {
            item.category.type === TYPE_INCOME ?
                currIncome += item.price :
                currOutcome += item.price
        })

        return (
            <>
                <header className="App-header">
                    <img src={logo} className="App-logo" alt="logo"/>
                    <Row>
                        <Col span={12}>
                            <MonthPicker
                                year={currentDate.year}
                                month={currentDate.month}
                                onChange={this.changeDate}
                            />
                        </Col>
                        <Col span={12}>
                            <TotalPrice
                                income={currIncome}
                                outcome={currOutcome}
                            />
                        </Col>
                    </Row>
                </header>


                <ViewTab
                    activeTab={tabView}
                    onTabViewChange={this.changeView}
                />

                <CreateBtn
                    onCreate={this.createItem}
                />

                {
                    tabView === LIST_VIEW &&
                    <PriceList
                        items={itemsWithCategory}
                        onEdit={this.editItem}
                        onDelete={this.deleteItem}
                    />
                }
                {
                    tabView === CHART_VIEW &&
                    <h1> Chart here</h1>
                }


            </>
        );
    }
}

export default Home;