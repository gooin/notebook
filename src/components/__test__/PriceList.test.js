import React from 'react';
import {shallow} from 'enzyme';
import PriceList from "../PriceList";


import {items, categries} from "../../containers/Home";
import toJson from 'enzyme-to-json';

const itemsWithCategory = items.map(item => {
    item.category = categries[item.cid]
    return item
})

const props = {
    items: itemsWithCategory,
    onEdit: jest.fn(),
    onDelete: jest.fn() // 模拟测试函数
}
let wrapper
describe('test PriceList component', () => {
    beforeEach(() => {
        wrapper = shallow(<PriceList {...props}/>)
    })

    it('match snapshot', () => {
        expect(toJson(wrapper)).toMatchSnapshot()
    })
    it('items has correct length', () => {
        expect(wrapper.find('Row').length).toEqual(itemsWithCategory.length)
        expect(wrapper.find('Button').length).toEqual(itemsWithCategory.length * 2)
    })
    it('render correct icon', () => {
        expect(wrapper.find('Icons').length).toEqual(itemsWithCategory.length)
        expect(wrapper.find('Icons').first().props().iconName).toEqual(itemsWithCategory[0].category.name)
    })

    it('should trigger the correct function callback', () => {
        // 模拟点击
        wrapper.find('Button').first().prop('onClick')()
        wrapper.find('Button').last().simulate('click')
        expect(props.onEdit).toHaveBeenCalledWith(itemsWithCategory[0])
        expect(props.onDelete).toHaveBeenCalled()
    })
})