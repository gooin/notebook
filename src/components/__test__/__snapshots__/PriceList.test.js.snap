// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test PriceList component match snapshot 1`] = `
<Fragment>
  <div
    key="1"
  >
    <Divider />
    <Row>
      <Col
        span={4}
      >
        <Icons
          iconName="food"
        />
      </Col>
      <Col
        span={8}
      >
        吃饭
      </Col>
      <Col
        span={4}
      >
        -
        200
      </Col>
      <Col
        span={6}
      >
        2020-11-12
      </Col>
      <Col
        span={1}
      >
        <Tooltip
          arrowPointAtCenter={false}
          autoAdjustOverflow={true}
          mouseEnterDelay={0.1}
          mouseLeaveDelay={0.1}
          placement="top"
          title="编辑"
          transitionName="zoom-big-fast"
        >
          <Button
            block={false}
            ghost={false}
            htmlType="button"
            icon={<ForwardRef(EditOutlined) />}
            loading={false}
            onClick={[Function]}
            shape="circle"
            size="small"
            type="primary"
          />
        </Tooltip>
      </Col>
      <Col
        span={1}
      >
        <Tooltip
          arrowPointAtCenter={false}
          autoAdjustOverflow={true}
          mouseEnterDelay={0.1}
          mouseLeaveDelay={0.1}
          placement="top"
          title="删除"
          transitionName="zoom-big-fast"
        >
          <Button
            block={false}
            danger={true}
            ghost={false}
            htmlType="button"
            icon={<ForwardRef(DeleteOutlined) />}
            loading={false}
            onClick={[Function]}
            shape="circle"
            size="small"
            type="primary"
          />
        </Tooltip>
      </Col>
    </Row>
  </div>
  <div
    key="2"
  >
    <Divider />
    <Row>
      <Col
        span={4}
      >
        <Icons
          iconName="salary"
        />
      </Col>
      <Col
        span={8}
      >
        假如发财了
      </Col>
      <Col
        span={4}
      >
        +
        200
      </Col>
      <Col
        span={6}
      >
        2020-11-12
      </Col>
      <Col
        span={1}
      >
        <Tooltip
          arrowPointAtCenter={false}
          autoAdjustOverflow={true}
          mouseEnterDelay={0.1}
          mouseLeaveDelay={0.1}
          placement="top"
          title="编辑"
          transitionName="zoom-big-fast"
        >
          <Button
            block={false}
            ghost={false}
            htmlType="button"
            icon={<ForwardRef(EditOutlined) />}
            loading={false}
            onClick={[Function]}
            shape="circle"
            size="small"
            type="primary"
          />
        </Tooltip>
      </Col>
      <Col
        span={1}
      >
        <Tooltip
          arrowPointAtCenter={false}
          autoAdjustOverflow={true}
          mouseEnterDelay={0.1}
          mouseLeaveDelay={0.1}
          placement="top"
          title="删除"
          transitionName="zoom-big-fast"
        >
          <Button
            block={false}
            danger={true}
            ghost={false}
            htmlType="button"
            icon={<ForwardRef(DeleteOutlined) />}
            loading={false}
            onClick={[Function]}
            shape="circle"
            size="small"
            type="primary"
          />
        </Tooltip>
      </Col>
    </Row>
  </div>
  <div
    key="2"
  >
    <Divider />
    <Row>
      <Col
        span={4}
      >
        <Icons
          iconName="travel"
        />
      </Col>
      <Col
        span={8}
      >
        去去医院体检
      </Col>
      <Col
        span={4}
      >
        -
        200
      </Col>
      <Col
        span={6}
      >
        2020-10-12
      </Col>
      <Col
        span={1}
      >
        <Tooltip
          arrowPointAtCenter={false}
          autoAdjustOverflow={true}
          mouseEnterDelay={0.1}
          mouseLeaveDelay={0.1}
          placement="top"
          title="编辑"
          transitionName="zoom-big-fast"
        >
          <Button
            block={false}
            ghost={false}
            htmlType="button"
            icon={<ForwardRef(EditOutlined) />}
            loading={false}
            onClick={[Function]}
            shape="circle"
            size="small"
            type="primary"
          />
        </Tooltip>
      </Col>
      <Col
        span={1}
      >
        <Tooltip
          arrowPointAtCenter={false}
          autoAdjustOverflow={true}
          mouseEnterDelay={0.1}
          mouseLeaveDelay={0.1}
          placement="top"
          title="删除"
          transitionName="zoom-big-fast"
        >
          <Button
            block={false}
            danger={true}
            ghost={false}
            htmlType="button"
            icon={<ForwardRef(DeleteOutlined) />}
            loading={false}
            onClick={[Function]}
            shape="circle"
            size="small"
            type="primary"
          />
        </Tooltip>
      </Col>
    </Row>
  </div>
  <div
    key="3"
  >
    <Divider />
    <Row>
      <Col
        span={4}
      >
        <Icons
          iconName="travel"
        />
      </Col>
      <Col
        span={8}
      >
        去去医院体检
      </Col>
      <Col
        span={4}
      >
        -
        200
      </Col>
      <Col
        span={6}
      >
        2020-9-12
      </Col>
      <Col
        span={1}
      >
        <Tooltip
          arrowPointAtCenter={false}
          autoAdjustOverflow={true}
          mouseEnterDelay={0.1}
          mouseLeaveDelay={0.1}
          placement="top"
          title="编辑"
          transitionName="zoom-big-fast"
        >
          <Button
            block={false}
            ghost={false}
            htmlType="button"
            icon={<ForwardRef(EditOutlined) />}
            loading={false}
            onClick={[Function]}
            shape="circle"
            size="small"
            type="primary"
          />
        </Tooltip>
      </Col>
      <Col
        span={1}
      >
        <Tooltip
          arrowPointAtCenter={false}
          autoAdjustOverflow={true}
          mouseEnterDelay={0.1}
          mouseLeaveDelay={0.1}
          placement="top"
          title="删除"
          transitionName="zoom-big-fast"
        >
          <Button
            block={false}
            danger={true}
            ghost={false}
            htmlType="button"
            icon={<ForwardRef(DeleteOutlined) />}
            loading={false}
            onClick={[Function]}
            shape="circle"
            size="small"
            type="primary"
          />
        </Tooltip>
      </Col>
    </Row>
  </div>
</Fragment>
`;
